libraries/custom_appbase/include/eosio/chain/application.hpp

```diff
@@ -52,23 +57,41 @@ class two_queue_executor {
    boost::asio::io_service& get_io_service() { return io_serv_; }
 
    bool execute_highest() {
-      if ( exec_window_ == exec_window::write ) {
-         // During write window only main thread is accessing anything in two_queue_executor, no locking required
-         if( !read_write_queue_.empty() && (read_only_queue_.empty() || *read_only_queue_.top() < *read_write_queue_.top()) )  {
-            // read_write_queue_'s top function's priority greater than read_only_queue_'s top function's, or read_only_queue_ empty
-            read_write_queue_.execute_highest();
-         } else if( !read_only_queue_.empty() ) {
-            read_only_queue_.execute_highest();
+      // execute for at least minimum runtime
+      const auto end = std::chrono::high_resolution_clock::now() + std::chrono::milliseconds(minimum_runtime_ms);
+
+      bool more = false;
+      while (true) {
+         if ( exec_window_ == exec_window::write ) {
+            // During write window only main thread is accessing anything in two_queue_executor, no locking required
+            if( !read_write_queue_.empty() && (read_only_queue_.empty() || *read_only_queue_.top() < *read_write_queue_.top()) )  {
+               // read_write_queue_'s top function's priority greater than read_only_queue_'s top function's, or read_only_queue_ empty
+               read_write_queue_.execute_highest();
+            } else if( !read_only_queue_.empty() ) {
+               read_only_queue_.execute_highest();
+            }
+            more = !read_only_queue_.empty() || !read_write_queue_.empty();
+         } else {
+            // When in read window, multiple threads including main app thread are accessing two_queue_executor, locking required
+            more = read_only_queue_.execute_highest_locked(false);
          }
-         return !read_only_queue_.empty() || !read_write_queue_.empty();
-      } else {
-         // When in read window, multiple threads including main app thread are accessing two_queue_executor, locking required
-         return read_only_queue_.execute_highest_locked(false);
+         if (!more || std::chrono::high_resolution_clock::now() > end)
+            break;
       }
+      return more;
    }
 
    bool execute_highest_read_only() {
-      return read_only_queue_.execute_highest_locked(true);
+      // execute for at least minimum runtime
+      const auto end = std::chrono::high_resolution_clock::now() + std::chrono::milliseconds(minimum_runtime_ms);
+
+      bool more = false;
+      while (true) {
+         more = read_only_queue_.execute_highest_locked( true );
+         if (!more || std::chrono::high_resolution_clock::now() > end)
+            break;
+      }
```

Comment:
Matter of preference of course!
```suggestion
       do {
         more = read_only_queue_.execute_highest_locked( true );
       } while (more && std::chrono::high_resolution_clock::now() < end);
```